src/hooks/execution/SpendingLimitHook.sol

```diff
@@ -0,0 +1,245 @@
+        }
+
+        // We will only filter based on functions that are known to use `msg.sender`.
+        // For signature-based approvals (e.g. permit), we can't do anything
+        // to guard, as anyone else can directly submit the calldata and the signature.
+        uint32 fnSel = uint32(bytes4(LibBytes.loadCalldata(data, 0x00)));
+        // `transfer(address,uint256)`.
+        if (fnSel == 0xa9059cbb) {
+            erc20s.p(to);
+            transferAmounts.p(LibBytes.loadCalldata(data, 0x24)); // `amount`.
+        }
+
+        // Sum transfer amounts, grouped by the ERC20s. In-place.
+        LibSort.groupSum(erc20s.data, transferAmounts.data);
+
+        // Collect the ERC20 balances before the batch execution.
+        for (uint256 i; i < erc20s.length(); ++i) {
+            address token = erc20s.getAddress(i);
+            balancesBefore.set(i, SafeTransferLib.balanceOf(token, msg.sender));
+        }
+
+        tempStorage.erc20s = erc20s.asAddressArray();
+        tempStorage.transferAmounts = transferAmounts.asUint256Array();
+        tempStorage.balancesBefore = balancesBefore;
+
+        return (IExecutionHook.beforeExecute.selector, abi.encode(tempStorage));
+    }
+
+    /// @inheritdoc IExecutionHook
+    function afterExecute(bytes32 keyHash, bool, bytes calldata, bytes calldata beforeExecuteData) external returns (bytes4) {
+        // Ensure spends in beforeExecuteData are within limits, revert if not.
+        TempStorage memory tempStorage = abi.decode(beforeExecuteData, (TempStorage));
+        SpendStorage storage spends = spendStorage[keyHash.hashSender(msg.sender)];
+
+        // Ensure spends in beforeExecuteData are within limits, revert if not.
+        // Perform after the `_execute`, so that in the case where `calls`
+        // contain a `setSpendLimit`, it will affect the `_incrementSpent`.
+
+        // `_incrementSpent` is an no-op if the token does not have an active spend limit.
+        _incrementSpent(spends.spends[address(0)], tempStorage.totalNativeSpend);
+
+        // Increments the spent amounts.
+        for (uint256 i; i < tempStorage.erc20s.length; ++i) {
+            address token = tempStorage.erc20s[i];
+            TokenSpendStorage storage tokenSpends = spends.spends[token];
+            if (tokenSpends.periods.length() == uint256(0)) continue;
+            _incrementSpent(
+                tokenSpends,
+                // While we can actually just use the difference before and after,
+                // we also want to let the sum of the transfer amounts in the calldata to be capped.
+                // This prevents tokens to be used as flash loans, and also handles cases
+                // where the actual token transfers might not match the calldata amounts.
+                // There is no strict definition on what constitutes spending,
+                // and we want to be as conservative as possible.
+                Math.max(
```

Comment:
this is just odd to me... what if you like flash loaned something? it would still count it as a spend here even though your balance stayed under some limit this would count it incorrectly no?